src/hotspot/cpu/s390/stubGenerator_s390.cpp

```diff
@@ -2857,6 +2858,44 @@ class StubGenerator: public StubCodeGenerator {
     return start;
   }
 
+  address generate_nmethod_entry_barrier() {
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "nmethod_entry_barrier");
+
+    address start = __ pc();
+
+    // Save caller's sp & return_pc
+    __ push_frame(frame::z_abi_16_size);
+    __ save_return_pc();
```

Comment:
Wrong order: return pc needs to get stored in the caller's frame header (before push_frame). See PPC64 or other usages on s390.

Reply:
Thanks.